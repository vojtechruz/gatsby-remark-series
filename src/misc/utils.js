import kebabCase from "lodash.kebabcase";
import defaultsDeep from "lodash.defaultsdeep";

/**
 * Default template.
 * @param {TemplateContext} context The context to be used.
 * @returns {string} The default template.
 */
export const DefaultTemplate = context => {
  let body;
  let title;

  if (context.pluginOptions.render.useLandingPage === true) {
    body = "";
    title = `<span>This post is part of the <a href="${context.slug}">${
      context.name
    }</a> series</span>`;
  } else {
    const list = context.items.reduce((code, item) => {
      const inner =
        item.draft !== true
          ? `<a href="${item.slug}">${item.title}</a>`
          : item.title;

      return `${code}<li>${inner}</li>`.trim();
    }, "");

    body = `<ol>${list}</ol>`;
    title = `<div>Other posts in the ${context.name} series</div>`;
  }

  return `
<div class="series-table-of-content">
  ${title}
  ${body}
</div>
  `.trim();
};

/**
 * Default options for the series.
 */
export const DefaultOptions = {
  resolvers: {
    slug: markdownNode => markdownNode.frontmatter.slug,
    date: markdownNode => markdownNode.frontmatter.date,
    draft: markdownNode => markdownNode.frontmatter.draft,
    order: markdownNode => markdownNode.frontmatter.order,
    series: markdownNode => markdownNode.frontmatter.series,
    toSlug: series => kebabCase(series)
  },
  render: {
    mode: "inline",
    placeholder: "bottom",
    template: DefaultTemplate,
    useLandingPage: false
  }
};

/**
 * Checks the type matches.
 * @param {string} type The expected type.
 * @param {string} name The name of the property.
 * @param {*} value The value of the property.
 */
function assertType(type, name, value) {
  if (typeof value !== type) {
    throw new Error(`${name} must be of type ${type}`);
  }
}

/**
 * Resolves the plug-in options to be used.
 * @param {PluginOptions} pluginOptions The user defined options.
 * @returns {PluginOptions} The resolved options.
 */
export function resolveOptions(pluginOptions) {
  const options = defaultsDeep({}, pluginOptions, DefaultOptions);

  assertType("string", "mode", options.render.mode);
  assertType("string", "placeholder", options.render.placeholder);
  assertType("function", "template", options.render.template);
  assertType("boolean", "useLandingPage", options.render.useLandingPage);

  options.render.mode = options.render.mode.toLowerCase();
  options.render.placeholder = options.render.placeholder.toLowerCase();

  assertType("function", "slug", options.resolvers.slug);
  assertType("function", "date", options.resolvers.date);
  assertType("function", "draft", options.resolvers.draft);
  assertType("function", "order", options.resolvers.order);
  assertType("function", "series", options.resolvers.series);
  assertType("function", "toSlug", options.resolvers.toSlug);

  if (options.render.useLandingPage === true) {
    assertType(
      "string",
      "landingPageComponent",
      options.render.landingPageComponent
    );

    if (options.render.landingPagePathPrefix != null) {
      assertType(
        "string",
        "landingPagePathPrefix",
        options.render.landingPagePathPrefix
      );
    }
  }

  return options;
}

/**
 * Resolves the series path (slug).
 * @param {string} name The name of the series.
 * @param {string} [pathPrefix] The global path prefix to be used.
 * @param {string} [pluginPathPrefix] The plugin defined path prefix to be used.
 * @returns {string} The resolved series path (slug).
 */
export function resolveSeriesPath(name, pathPrefix, pluginPathPrefix) {
  const path = [];

  assertType("string", "name", name);

  if (typeof pathPrefix === "string") {
    const cleanPathPrefix = pathPrefix.replace(/^\/|\/$/g, "");
    if (cleanPathPrefix.trim().length > 0) {
      path.push(cleanPathPrefix);
    }
  }

  if (typeof pluginPathPrefix === "string") {
    const cleanpluginPathPrefix = pluginPathPrefix.replace(/^\/|\/$/g, "");
    if (cleanpluginPathPrefix.trim().length > 0) {
      path.push(cleanpluginPathPrefix);
    }
  }

  path.push(name.replace(/^\/|\/$/g, ""));

  return `/${path.join("/")}`;
}

/**
 * Sorts the items in a series.
 * Items are sorted by order, falling back to date then title.
 * @param {SeriesItem} left The left item to be sorted.
 * @param {SeriesItem} right The right item to be sorted.
 * @returns {number} The sorted resolution.
 */
export function sortItems(left, right) {
  const orderLeft = left.order !== null ? Number(left.order) : NaN;
  const orderRight = right.order !== null ? Number(right.order) : NaN;

  if (isNaN(orderLeft) || isNaN(orderRight)) {
    if (!isNaN(orderLeft)) return -1;
    if (!isNaN(orderRight)) return 1;
  } else if (orderLeft !== orderRight) {
    return orderLeft - orderRight;
  }

  const dateLeft =
    left.date !== null ? new Date(left.date) : new Date(undefined);
  const dateRight =
    right.date !== null ? new Date(right.date) : new Date(undefined);

  if (isNaN(dateLeft.getTime()) || isNaN(dateRight.getTime())) {
    if (!isNaN(dateLeft.getTime())) return -1;
    if (!isNaN(dateRight.getTime())) return 1;
  } else if (dateLeft.getTime() !== dateRight.getTime()) {
    return dateLeft.getTime() - dateRight.getTime();
  }

  if (left.title == null || right.title == null) {
    if (left.title != null) return -1;
    if (right.title != null) return 1;
    return 0;
  }

  return left.title.localeCompare(right.title);
}
