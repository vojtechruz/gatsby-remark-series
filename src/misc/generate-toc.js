import { resolveSeriesPath } from "./utils";

const localCache = {};

/**
 * Validates the context of the request.
 * @param {GatsbyContext} context the gatsby context to be used.
 * @param {PluginOptions} options the plugin options to be used.
 * @returns {boolean} true if the context is valid; false otherwise.
 */
function validContext(context, options) {
  const series = options.series(context.markdownNode);
  if (series == null) return false;

  return true;
}

/**
 * Renders the table of content at the top of the page.
 * @param {object} context the template context to be used.
 * @returns {void}
 */
function renderTop({ options, ...context }) {
  if (
    options.render.placeholder === "top" ||
    options.render.placeholder === "both"
  ) {
    context.markdownAST.children.unshift({
      type: "html",
      value: options.template({
        placeholder: "top",
        name: context.name,
        slug: context.slug,
        items: context.items,
        pluginOptions: options,
        markdownNode: context.markdownNode
      })
    });
  }
}

/**
 * Renders the table of content at the bottom of the page.
 * @param {object} context the template values to be used.
 * @returns {void}
 */
function renderBottom({ options, ...context }) {
  if (
    options.render.placeholder === "bottom" ||
    options.render.placeholder === "both"
  ) {
    context.markdownAST.children.push({
      type: "html",
      value: options.render.template({
        placeholder: "bottom",
        name: context.name,
        slug: context.slug,
        items: context.items,
        pluginOptions: options,
        markdownNode: context.markdownNode
      })
    });
  }
}

/**
 * Gets all nodes associated to a series.
 * @param {GatsbyContext} context the gatsby context to be used.
 * @param {PluginOptions} options the plugin options to be used.
 * @returns {Array<object>} the nodes in the series.
 */
function getSeriesItems(context, options) {
  const series = options.series(context.markdownNode);
  const cacheKey = context.createContentDigest(series);

  // cache gets cleared on every build, but that should be enough
  // to prevent siblings from unnecessarily rebuilding the list.
  let items = localCache[cacheKey];
  if (items == null) {
    items = context
      .getNodes()

      // gets markdown remark elements that have a series
      .filter(
        node =>
          node.internal.type === "MarkdownRemark" &&
          options.series(node) === series
      )

      // map to something meaningful
      .map(node => {
        return {
          title: node.frontmatter.title,
          slug: options.slug(node),
          date: options.date(node),
          draft: options.draft(node),
          order: options.order(node),
          series: options.series(node)
        };
      })

      // sort
      // 1. order if available
      // 2. date if available
      // 3. title
      .sort((left, right) => {
        const orderLeft = Number(left.order);
        const orderRight = Number(right.order);

        if (!isNaN(orderLeft) && !isNaN(orderRight)) {
          return orderLeft - orderRight;
        }

        if (!isNaN(orderLeft)) return -1;
        if (!isNaN(orderRight)) return 1;

        const dateLeft = new Date(left.date);
        const dateRight = new Date(right.date);

        if (!isNaN(dateLeft.getTime()) && !isNaN(dateRight.getTime())) {
          return dateLeft - dateRight;
        }

        if (!isNaN(dateLeft.getTime())) return -1;
        if (!isNaN(dateRight.getTime())) return 1;

        return left.title.localCompare(right.title);
      });

    localCache[cacheKey] = items;
  }

  return items;
}

/**
 * Renders the table of content.
 * @param {GatsbyContext} context the gatsby context to be used.
 * @param {PluginOptions} options the plugin options to be used.
 * @returns {AST} the modified markdownAST.
 */
export default (context, options) => {
  if (!validContext(context, options)) return context.markdownAST;

  const { markdownAST, markdownNode, pathPrefix } = context;

  const series = options.series(markdownNode);
  const items = getSeriesItems(context, options);

  const path = resolveSeriesPath(
    options.toSlug(series),
    pathPrefix,
    options.pathPrefix
  );

  renderTop({ options, items, name: series, slug: path, ...context });
  renderBottom({ options, items, name: series, slug: path, ...context });

  return markdownAST;
};
