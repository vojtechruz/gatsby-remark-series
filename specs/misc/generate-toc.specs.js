import { readFile } from "fs";
import { promisify } from "util";

import Remark from "remark";
import grayMatter from "gray-matter";
import defaultsDeep from "lodash.defaultsdeep";

import generateTOC from "../../src/misc/generate-toc";
import { DefaultOptions } from "../../src/misc/utils";

describe("generate-toc", () => {
  const remark = new Remark().data("settings", {
    commonmark: true,
    footnotes: true,
    pedantic: true
  });

  const createContentDigest = () => {
    return Math.random()
      .toString(36)
      .substring(7);
  };

  const getNodes = () => {
    return [
      {
        internal: {
          type: "MarkdownRemark"
        },
        frontmatter: {
          title: "My Title 2",
          date: "2019-01-01",
          series: "My Title Series"
        }
      },
      { internal: { type: "other" } },
      {
        internal: {
          type: "MarkdownRemark"
        },
        frontmatter: {
          title: "My Title 3",
          date: "2019-02-01",
          series: "My Title Series"
        }
      },
      {
        internal: {
          type: "MarkdownRemark"
        },
        frontmatter: {
          title: "My Title",
          date: "2019-01-01",
          series: "Other Series"
        }
      },
      { internal: { type: "other" } }
    ];
  };

  const readFilePromise = promisify(readFile);

  const getSeriesMarkdown = async (options, pathPrefix) => {
    const seriesMarkdown = await readFilePromise(
      "./specs/files/series.md",
      "utf-8"
    );

    const markdownAST = remark.parse(seriesMarkdown.trim());
    const countPrePlugin = markdownAST.children.length;

    const postMarkdownAST = generateTOC(
      {
        createContentDigest,
        getNodes,
        markdownAST,
        markdownNode: { frontmatter: grayMatter(seriesMarkdown.trim()).data },
        pathPrefix
      },
      options
    );

    return [postMarkdownAST, countPrePlugin];
  };

  test("returns the exact AST when no series is detected", async () => {
    const noSeriesMarkdown = await readFilePromise(
      "./specs/files/no-series.md",
      "utf-8"
    );

    const markdownAST = remark.parse(noSeriesMarkdown.trim());
    const countPrePlugin = markdownAST.children.length;

    const postMarkdownAST = generateTOC(
      {
        createContentDigest,
        markdownAST,
        markdownNode: { frontmatter: grayMatter(noSeriesMarkdown.trim()).data }
      },
      DefaultOptions
    );

    expect(countPrePlugin).toBe(postMarkdownAST.children.length);
    expect(postMarkdownAST).toMatchSnapshot();
  });

  test("returns a modified version of the AST when series is detected", async () => {
    const [postMarkdownAST, countPrePlugin] = await getSeriesMarkdown(
      DefaultOptions
    );

    expect(countPrePlugin).not.toBe(postMarkdownAST.children.length);
    expect(postMarkdownAST).toMatchSnapshot();
  });

  test("adds a children on top of the AST's children when top placeholder is used", async () => {
    const [postMarkdownAST] = await getSeriesMarkdown(
      defaultsDeep({ render: { placeholder: "top" } }, DefaultOptions)
    );

    expect(postMarkdownAST).toMatchSnapshot();
    expect(postMarkdownAST.children[0].type).toBe("html");
    expect(postMarkdownAST.children[0].value).toContain(
      "series-table-of-content"
    );
  });

  test("adds a children on bottom of the AST's children when bottom placeholder is used", async () => {
    const [postMarkdownAST] = await getSeriesMarkdown(
      defaultsDeep({ render: { placeholder: "bottom" } }, DefaultOptions)
    );

    expect(postMarkdownAST).toMatchSnapshot();
    expect(postMarkdownAST.children[4].type).toBe("html");
    expect(postMarkdownAST.children[4].value).toContain(
      "series-table-of-content"
    );
  });

  test("renders inline toc references", async () => {
    const [postMarkdownAST] = await getSeriesMarkdown(
      defaultsDeep({ render: { useLandingPage: false } }, DefaultOptions)
    );

    expect(postMarkdownAST).toMatchSnapshot();
    expect(postMarkdownAST.children[4].type).toBe("html");
    expect(postMarkdownAST.children[4].value).toContain(
      "Other posts in the My Title Series series"
    );
  });

  test("renders reference to external toc", async () => {
    const [postMarkdownAST] = await getSeriesMarkdown(
      defaultsDeep({ render: { useLandingPage: true } }, DefaultOptions)
    );

    expect(postMarkdownAST).toMatchSnapshot();
    expect(postMarkdownAST.children[4].type).toBe("html");
    expect(postMarkdownAST.children[4].value).toContain(
      'This post is part of the <a href="/my-title-series">My Title Series</a>'
    );
  });

  test("uses the pluginPathPrefix in external toc", async () => {
    const [postMarkdownAST] = await getSeriesMarkdown(
      defaultsDeep(
        {
          render: {
            useLandingPage: true,
            landingPagePathPrefix: "series"
          }
        },
        DefaultOptions
      )
    );

    expect(postMarkdownAST).toMatchSnapshot();
    expect(postMarkdownAST.children[4].type).toBe("html");
    expect(postMarkdownAST.children[4].value).toContain(
      'This post is part of the <a href="/series/my-title-series">My Title Series</a>'
    );
  });

  test("uses the pathPrefix in external toc", async () => {
    const [postMarkdownAST] = await getSeriesMarkdown(
      defaultsDeep(
        {
          render: {
            useLandingPage: true,
            landingPagePathPrefix: "series"
          }
        },
        DefaultOptions
      ),
      "blog"
    );

    expect(postMarkdownAST).toMatchSnapshot();
    expect(postMarkdownAST.children[4].type).toBe("html");
    expect(postMarkdownAST.children[4].value).toContain(
      'This post is part of the <a href="/blog/series/my-title-series">My Title Series</a>'
    );
  });
});
