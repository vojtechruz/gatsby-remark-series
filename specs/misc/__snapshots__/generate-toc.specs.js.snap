// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-toc adds a children on bottom of the AST's children when bottom placeholder is used 1`] = `
Object {
  "children": Array [
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 1,
          "offset": 3,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 26,
              "line": 4,
              "offset": 66,
            },
            "indent": Array [
              1,
              1,
            ],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 4,
            },
          },
          "type": "text",
          "value": "title: \\"My Title\\"
date: \\"2019-03-18\\"
series: \\"My Title Series\\"",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 26,
          "line": 4,
          "offset": 66,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 4,
        },
      },
      "type": "paragraph",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 5,
          "offset": 70,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 67,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 7,
              "offset": 82,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 7,
              "offset": 74,
            },
          },
          "type": "text",
          "value": "My Title",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 7,
          "offset": 82,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 7,
          "offset": 72,
        },
      },
      "type": "heading",
    },
    Object {
      "type": "html",
      "value": "<div class=\\"series-table-of-content\\">
  <div>Other posts in the My Title Series series</div>
  <ol><li><a href=\\"undefined\\">My Title 2</a></li><li><a href=\\"undefined\\">My Title 3</a></li></ol>
</div>",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 7,
      "offset": 82,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`generate-toc adds a children on top of the AST's children when top placeholder is used 1`] = `
Object {
  "children": Array [
    Object {
      "type": "html",
      "value": "<div class=\\"series-table-of-content\\">
  <div>Other posts in the My Title Series series</div>
  <ol><li><a href=\\"undefined\\">My Title 2</a></li><li><a href=\\"undefined\\">My Title 3</a></li></ol>
</div>",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 1,
          "offset": 3,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 26,
              "line": 4,
              "offset": 66,
            },
            "indent": Array [
              1,
              1,
            ],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 4,
            },
          },
          "type": "text",
          "value": "title: \\"My Title\\"
date: \\"2019-03-18\\"
series: \\"My Title Series\\"",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 26,
          "line": 4,
          "offset": 66,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 4,
        },
      },
      "type": "paragraph",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 5,
          "offset": 70,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 67,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 7,
              "offset": 82,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 7,
              "offset": 74,
            },
          },
          "type": "text",
          "value": "My Title",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 7,
          "offset": 82,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 7,
          "offset": 72,
        },
      },
      "type": "heading",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 7,
      "offset": 82,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`generate-toc renders inline toc references 1`] = `
Object {
  "children": Array [
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 1,
          "offset": 3,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 26,
              "line": 4,
              "offset": 66,
            },
            "indent": Array [
              1,
              1,
            ],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 4,
            },
          },
          "type": "text",
          "value": "title: \\"My Title\\"
date: \\"2019-03-18\\"
series: \\"My Title Series\\"",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 26,
          "line": 4,
          "offset": 66,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 4,
        },
      },
      "type": "paragraph",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 5,
          "offset": 70,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 67,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 7,
              "offset": 82,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 7,
              "offset": 74,
            },
          },
          "type": "text",
          "value": "My Title",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 7,
          "offset": 82,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 7,
          "offset": 72,
        },
      },
      "type": "heading",
    },
    Object {
      "type": "html",
      "value": "<div class=\\"series-table-of-content\\">
  <div>Other posts in the My Title Series series</div>
  <ol><li><a href=\\"undefined\\">My Title 2</a></li><li><a href=\\"undefined\\">My Title 3</a></li></ol>
</div>",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 7,
      "offset": 82,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`generate-toc renders reference to external toc 1`] = `
Object {
  "children": Array [
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 1,
          "offset": 3,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 26,
              "line": 4,
              "offset": 66,
            },
            "indent": Array [
              1,
              1,
            ],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 4,
            },
          },
          "type": "text",
          "value": "title: \\"My Title\\"
date: \\"2019-03-18\\"
series: \\"My Title Series\\"",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 26,
          "line": 4,
          "offset": 66,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 4,
        },
      },
      "type": "paragraph",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 5,
          "offset": 70,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 67,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 7,
              "offset": 82,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 7,
              "offset": 74,
            },
          },
          "type": "text",
          "value": "My Title",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 7,
          "offset": 82,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 7,
          "offset": 72,
        },
      },
      "type": "heading",
    },
    Object {
      "type": "html",
      "value": "<div class=\\"series-table-of-content\\">
  <span>This post is part of the <a href=\\"/my-title-series\\">My Title Series</a> series</span>
  
</div>",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 7,
      "offset": 82,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`generate-toc returns a modified version of the AST when series is detected 1`] = `
Object {
  "children": Array [
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 1,
          "offset": 3,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 26,
              "line": 4,
              "offset": 66,
            },
            "indent": Array [
              1,
              1,
            ],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 4,
            },
          },
          "type": "text",
          "value": "title: \\"My Title\\"
date: \\"2019-03-18\\"
series: \\"My Title Series\\"",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 26,
          "line": 4,
          "offset": 66,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 4,
        },
      },
      "type": "paragraph",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 5,
          "offset": 70,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 67,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 7,
              "offset": 82,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 7,
              "offset": 74,
            },
          },
          "type": "text",
          "value": "My Title",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 7,
          "offset": 82,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 7,
          "offset": 72,
        },
      },
      "type": "heading",
    },
    Object {
      "type": "html",
      "value": "<div class=\\"series-table-of-content\\">
  <div>Other posts in the My Title Series series</div>
  <ol><li><a href=\\"undefined\\">My Title 2</a></li><li><a href=\\"undefined\\">My Title 3</a></li></ol>
</div>",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 7,
      "offset": 82,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`generate-toc returns the exact AST when no series is detected 1`] = `
Object {
  "children": Array [
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 1,
          "offset": 3,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 19,
              "line": 3,
              "offset": 40,
            },
            "indent": Array [
              1,
            ],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 4,
            },
          },
          "type": "text",
          "value": "title: \\"My Title\\"
date: \\"2019-03-18\\"",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 19,
          "line": 3,
          "offset": 40,
        },
        "indent": Array [
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 4,
        },
      },
      "type": "paragraph",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 4,
          "offset": 44,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 4,
          "offset": 41,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 6,
              "offset": 56,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 6,
              "offset": 48,
            },
          },
          "type": "text",
          "value": "My Title",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 6,
          "offset": 56,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 6,
          "offset": 46,
        },
      },
      "type": "heading",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 6,
      "offset": 56,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`generate-toc uses the pathPrefix in external toc 1`] = `
Object {
  "children": Array [
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 1,
          "offset": 3,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 26,
              "line": 4,
              "offset": 66,
            },
            "indent": Array [
              1,
              1,
            ],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 4,
            },
          },
          "type": "text",
          "value": "title: \\"My Title\\"
date: \\"2019-03-18\\"
series: \\"My Title Series\\"",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 26,
          "line": 4,
          "offset": 66,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 4,
        },
      },
      "type": "paragraph",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 5,
          "offset": 70,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 67,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 7,
              "offset": 82,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 7,
              "offset": 74,
            },
          },
          "type": "text",
          "value": "My Title",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 7,
          "offset": 82,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 7,
          "offset": 72,
        },
      },
      "type": "heading",
    },
    Object {
      "type": "html",
      "value": "<div class=\\"series-table-of-content\\">
  <span>This post is part of the <a href=\\"/blog/series/my-title-series\\">My Title Series</a> series</span>
  
</div>",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 7,
      "offset": 82,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`generate-toc uses the pluginPathPrefix in external toc 1`] = `
Object {
  "children": Array [
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 1,
          "offset": 3,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 26,
              "line": 4,
              "offset": 66,
            },
            "indent": Array [
              1,
              1,
            ],
            "start": Object {
              "column": 1,
              "line": 2,
              "offset": 4,
            },
          },
          "type": "text",
          "value": "title: \\"My Title\\"
date: \\"2019-03-18\\"
series: \\"My Title Series\\"",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 26,
          "line": 4,
          "offset": 66,
        },
        "indent": Array [
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 2,
          "offset": 4,
        },
      },
      "type": "paragraph",
    },
    Object {
      "position": Position {
        "end": Object {
          "column": 4,
          "line": 5,
          "offset": 70,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 67,
        },
      },
      "type": "thematicBreak",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 7,
              "offset": 82,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 7,
              "offset": 74,
            },
          },
          "type": "text",
          "value": "My Title",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 7,
          "offset": 82,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 7,
          "offset": 72,
        },
      },
      "type": "heading",
    },
    Object {
      "type": "html",
      "value": "<div class=\\"series-table-of-content\\">
  <span>This post is part of the <a href=\\"/series/my-title-series\\">My Title Series</a> series</span>
  
</div>",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 7,
      "offset": 82,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
